### ISSUE TITLE
Inner test cases in loops

### ISSUE THREAD

**Issue by toby3d** [bitwes/Gut#193]
After writing several tests, I found that I can glue them into one function with several typical cases in `for` iterations. But I did not quite understand how I can indicate that each iteration is a separate inner test (maybe even apply some kind of proper name to it) without using subclasses.

### Tested class
```gdscript
class_name Player
extends Node

signal balance_changed(old_balance, new_balance)

var money: int = 0 setget set_money

func set_money(value: int) -> void:
	if value == money:
		return

	emit_signal("balance_changed", money, value)

	money = value
```

### Tests script
```gdscript
extends "res://addons/gut/test.gd"

func test_money() -> void:
	for test_case in [{
		"name": "balance not changed, don't do anything",
		"old_balance": 420,
		"new_balance": 420,
		"result": 420,
		"must_emit": false,
	},{
		"name": "take money from player",
		"old_balance": 420,
		"new_balance": 240,
		"result": 240,
		"must_emit": true,
	},{
		"name": "give money to player",
		"old_balance": 240,
		"new_balance": 420,
		"result": 420,
		"must_emit": true,
	}]:
		# NOTE(toby3d): How I can set each loop iteration here as inner test case?
		# TODO(toby3d): Use test_case.name as inner test name.

		var player: Player = Player.new()
		var old_balance: int
		var new_balance: int

		# NOTE(toby3d): I can't use assert_accessors method, since I do not have get_money getter.

		if test_case.has("old_balance"):
			player.money = test_case.old_balance
			old_balance = test_case.old_balance

		watch_signals(player)

		if test_case.has("new_balance"):
			player.money = test_case.new_balance
			new_balance = test_case.new_balance

		if test_case.has("result"):
			assert_eq(player.money, test_case.result)

		if not test_case.has("must_emit") or !test_case.must_emit:
			assert_signal_not_emitted(player, "balance_changed")
			continue

		assert_signal_emitted_with_parameters(player, "balance_changed", [old_balance, new_balance])
```

**Comment by bitwes** [comment]
This isn't the best approach for writing unit tests.  Unit tests should be simple and easy to understand.  This sometimes results in duplicate code which can feel wrong.  We have a tendency to not want to duplicate code, which is good for your program but bad for unit tests (in general).  For unit tests though, simplicity and readability are more important than some duplicate code.  This also will lead to hard to read output when something fails since the exact scenario that failed is hard to determine from the output.  

I think you had it right before you tried to glue these together.  Each of your `if` statements should be a test function.  This way you can have descriptive names so the output is easier to understand and each test case is easy to understand.

In GUT 7.0.0 you could use parameterized tests to accomplish this.  You could group the 2nd and 3rd test cases together into a single test that took parameters.  You shouldn't group in the first because it has a different assertion (the signal should not be emitted).  I would advise against using parameterized tests for only 2 cases.  You are better off making two tests with descriptive names so that your output is easier to understand when something fails.

**Comment by bitwes** [comment]
I hit "comment" too soon.  One thing you can do that might improve readability would be to use Inner Test classes to group your tests together.  You can then use the `before` and `after` methods to have some base setup that is common to a few tests.

I assume the code you posted was for the a `test_player` script.
```
extends "res://addons/gut/test.gd"

class TestProperties:
  extends "res://addons/gut/test.gd"
  
  func test_prop1():
    pass
  
  func test_prop2():
    pass

class TestMoney
  extends "res://addons/gut/test.gd"

  var _player = null
  
  func before_each():
    _player = Player.new()

  func test_when_balance_not_changed_signal_is_not_emitted():
    ...
  
  func test_when_player_loses_money_then_balance_changed_emitted():
    ...

  func test_when_player_gains_money_then_balance_changed_emitted():
    ...

```

**Comment by toby3d** [comment]
> This isn't the best approach for writing unit tests. Unit tests should be simple and easy to understand. This sometimes results in duplicate code which can feel wrong. We have a tendency to not want to duplicate code, which is good for your program but bad for unit tests (in general). For unit tests simplicity and readability are more important than some duplicate code. This also will lead to hard to read output when something fails since the exact scenario that failed is hard to determine from the output.

I understand why it's better to keep the tests separate, even if they're mostly duplicate. I may not have picked up a very good example, but the point is that sometimes there are tests that change at least two parameters (input + output) and have to be tested over 2 different cases.

Suppose I simplify an example to a certain class that can only accept a `String` and returns a certain `int` in a known range (aka `Enum`). The test will be the same: `assert_eq(X, exampleClass.doSomething(Y))`. But there are 15 behavioral scenarios, suppose. In this case, it is easier for me to apply `for`. The readability doesn't get worse, because the test body is obviously the same and the cases themselves describe themselves ("I give it", "expect that"). Plus, I can add and remove cases respectively by deleting/adding only 2-3 lines. Gut will already run these tests successfully, but it will register them as one. And this already disrupts the readability, as it is misleading. 

I don't criticize subclasses as a way of dividing tests into groups, I don't have a problem with that. Probably, it is technically difficult to implement some method `sub_test("test name", inner_func)` in GDScript (Godot doesn't support storing functions in variables, right?). I just wanted to clarify if there is a way to separate tests in loops, maybe I just missed something.

**Comment by bitwes** [comment]
Checkout Parameterized Tests then.  This is the formal way of doing it.  You can make a single test that takes parameters and gets called N times with different values passed in.  This is new 7.0.0 so if you aren't using that version you'll have to upgrade.  Checkout the release notes b/c there are some minor breaking changes.  

https://github.com/bitwes/Gut/releases/tag/v7.0.0

https://github.com/bitwes/Gut/wiki/Parameterized-Tests

If for some reason parameterized tests don't work for you, then a better approach would be to create a utility method that does an assert based on parameters and then just call it N times directly.

```
func _assert_money(a, b, c)
  a.set_b(b)
  a.add_money(c)
  assert_eq(a.get_money(), a + b)

func test_monies()
  _assert_money(player, 1, 2)
  _assert_money(player, 2, 3)
  etc
```
Putting the `for` loop inside the test is hard to follow.  With Parameterized Tests you get the `for` loop for free through GUT.

**Comment by toby3d** [comment]
Okay, thanks! :+1:
